src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
-        let tree_height = tree_height(tree_size);
-        assert!(tree_height as u8 <= Self::HEIGHT);
-
-        // Initialize the merkle tree.
-        let mut tree = Vec::with_capacity(tree_size);
-        let empty_hash = hash_empty::<P::H>(&parameters)?;
-        for _ in 0..tree_size {
-            tree.push(empty_hash.clone());
-        }
-
-        // Compute the starting indices for each level of the tree.
+    /// Returns a new merkle tree. `leaves.len()` should be power of two.
+    pub fn new<L: ToBytes>(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        leaves: &[L],
+    ) -> Result<Self, crate::Error> {
+        let leaf_nodes_size = leaves.len(); // size of the leaf layer
+        assert!(
+            leaf_nodes_size.is_power_of_two(),
+            "`leaves.len() should be power of two"
+        );
+        let non_leaf_nodes_size = leaf_nodes_size - 1;
+
+        let tree_height = tree_height(non_leaf_nodes_size + leaf_nodes_size);
+
+        let hash_of_empty: TwoToOneDigest<P> = P::TwoToOneHash::evaluate_both(
+            two_to_one_hash_param,
+            &vec![0u8; P::TwoToOneHash::LEFT_INPUT_SIZE_BITS / 8],
+            &vec![0u8; P::TwoToOneHash::RIGHT_INPUT_SIZE_BITS / 8],
+        )?;
+
+        // initialize the merkle tree as array of nodes in level order
+        let mut non_leaf_nodes: Vec<TwoToOneDigest<P>> = (0..non_leaf_nodes_size)
+            .map(|_| hash_of_empty.clone())
+            .collect();
+        let mut leaf_nodes: Vec<LeafDigest<P>> = Vec::with_capacity(leaf_nodes_size);
+
+        // Compute the starting indices for each non-leaf level of the tree
         let mut index = 0;
-        let mut level_indices = Vec::with_capacity(tree_height);
-        for _ in 0..tree_height {
+        let mut level_indices = Vec::with_capacity(tree_height - 1);
+        for _ in 0..(tree_height - 1) {
             level_indices.push(index);
             index = left_child(index);
         }
 
-        // Compute and store the hash values for each leaf.
-        let last_level_index = level_indices.pop().unwrap_or(0);
-        let mut buffer = [0u8; 128];
-        for (i, leaf) in leaves.iter().enumerate() {
-            tree[last_level_index + i] = hash_leaf::<P::H, _>(&parameters, leaf, &mut buffer)?;
+        // compute and store hash values for each leaf
+        for leaf in leaves.iter() {
+            leaf_nodes.push(P::LeafHash::evaluate(
+                leaf_hash_param,
+                &ark_ff::to_bytes!(leaf)?,
+            )?)
+        }
+
+        // compute the hash values for the non-leaf bottom layer
+        {
+            let start_index = level_indices.pop().unwrap();
+            let upper_bound = left_child(start_index);
+            for current_index in start_index..upper_bound {
+                let left_leaf_index = left_child(current_index) - upper_bound;
+                let right_leaf_index = right_child(current_index) - upper_bound;
```

Comment:
I'm a bit confused as to how the indexing works if we need to subtract the upperbound from each index. Is the indexing not that the `i`th nodes children are at `2*(1 + i)` and `2(1 + i) + 1`?

Reply:
`left_child(current_index)` returns a position of leaf with respect to the whole tree in level order, and `left_leaf_index` represents the position of leaf in the array of leaves, so `-upper_bound` does the conversion. 

For example, we have
```
       Root
      /     \
    A       B
  /   \    /   \
C     D    E   F
```
The tree in level order is [Root, A, B, C, D, E, F], whereas the leaves array is [C,D,E,F]. Suppose we want to get left child of `B`. Then `left_child(2)` returns `2*2+1 = 5`, which is `E`'s position in the level order array. However, we want to get the position in leaves array, so we offset it by `upperbound=left_child(1) = 3` and `5-3=2`? Does that make sense?

Reply:
Oh I see, this does make sense. Can there be a comment explaining that it has to be shifted in order to get the position in the leaf index?

Reply:
sure. I will add this comment soon